src/sentry/reprocessing2.py

```diff
@@ -347,73 +351,98 @@ def _get_info_reprocessed_key(group_id):
-    event IDs in Redis. We need this because Snuba cannot handle many tiny messages and prefers big ones instead.
+    event IDs in Redis. We need this because Snuba cannot handle many tiny
+    messages and prefers big ones instead.
 
-    Best performance is achieved when timestamps are close together. Luckily we
-    happen to iterate through events ordered by timestamp.
+    For optimal performance, the datetimes should be close to each other. This
+    "soft" precondition is fulfilled in `reprocess_group` by iterating through
+    events in timestamp order.
 
     Ideally we'd have batching implemented via a service like buffers, but for
     more than counters.
     """
 
     client = _get_sync_redis_client()
-    key = f"re2:remaining:{project_id}:{old_group_id}"
+    # We explicitly cluster by only project_id and group_id here such that our
+    # RENAME command later succeeds.
+    #
+    # We also use legacy string formatting here because new-style Python
+    # formatting is quite confusing when the output string is supposed to
+    # contain {}.
+    key = f"re2:remaining:{{{project_id}:{old_group_id}}}"
+
+    llen = None
 
     if datetime_to_event:
-        client.lpush(
+        llen = client.lpush(
             key,
             *(f"{to_timestamp(datetime)};{event_id}" for datetime, event_id in datetime_to_event),
         )
         client.expire(key, settings.SENTRY_REPROCESSING_SYNC_TTL)
 
-    if (
-        force_flush_batch
-        or client.llen(key) > settings.SENTRY_REPROCESSING_REMAINING_EVENTS_BUF_SIZE
-    ):
-        event_ids_batch = []
-        min_datetime = None
-        max_datetime = None
-
-        # TODO: Redis 6 introduces LPOP with <count> argument, use here?
-        while True:
-            row = client.lpop(key)
-            if row is None:
-                break
+    MAX_LLEN = settings.SENTRY_REPROCESSING_REMAINING_EVENTS_BUF_SIZE
 
-            datetime_raw, event_id = row.split(";")
-            datetime = to_datetime(float(datetime_raw))
+    if force_flush_batch or (llen is not None and llen > MAX_LLEN) or client.llen(key) > MAX_LLEN:
+        new_key = f"{key}:{uuid.uuid4().hex}"
 
-            assert datetime is not None
-
-            if min_datetime is None or datetime < min_datetime:
-                min_datetime = datetime
-            if max_datetime is None or datetime > max_datetime:
-                max_datetime = datetime
-
-            event_ids_batch.append(event_id)
+        try:
+            # Rename `key` to a new temp key that is passed to celery task. We
+            # use `renamenx` instead of `rename` only to detect UUID collisions.
+            assert client.renamenx(key, new_key), "UUID collision for new_key?"
+        except redis.exceptions.ResponseError:
+            # `key` does not exist in Redis. `ResponseError` is a bit too broad
+            # but it seems we'd have to do string matching on error message
+            # otherwise.
+            return
```

Comment:
Should we log anything here?

Reply:
this condition happens "too often", i.e. when reprocessing is finished and we execute force-flushing